fflib/src/classes/fflib_QueryFactory.cls

```diff
@@ -108,7 +108,22 @@ public class fflib_QueryFactory { //No explicit sharing declaration - inherit fr
 			}
 
 			if(token != null && i.hasNext() && tokenDescribe.getSOAPType() == Schema.SOAPType.ID){
-				lastSObjectType = tokenDescribe.getReferenceTo()[0]; //if it's polymorphic doesn't matter which one we get
+				if(relatedSObjectType == null) {
+					lastSObjectType = tokenDescribe.getReferenceTo()[0]; //if it's polymorphic get the first one - user did not specify the one to use.
+				}else{
+					List<Schema.sObjectType> relatedObjs = tokenDescribe.getReferenceTo(); //if it's polymorphic, it matters which one we use - i.e. Lead.Owner is GROUP|USER and each has different fields.
+					if(relatedObjs.size() > 1) {
+						String relatedSObjectName = fflib_SObjectDescribe.getDescribe(relatedSObjectType).getDescribe().getName();
+						for(Schema.sObjectType sot : relatedObjs) {
+							if(fflib_SObjectDescribe.getDescribe(sot).getDescribe().getName() == relatedSObjectName){
```

Comment:
Out of interest why are we comparing the names and not just the sobject type objects?